### ISSUE TITLE
AccountRefresh deactivating Account

### ISSUE THREAD

**Issue by eren-kemer** [rmbolger/Posh-ACME#472]
Heya! First of all, thank you for the module!

I am trying to use Posh-ACME with [Secardeo certACME](https://www.secardeo.com/products/certacme/) using one account per hostname (with only **one** Posh-ACME client)
So currently i have a workflow that creates a new account for the hostname if it doesn't exist and then requests the certificate.
If it does exist, it uses `Set-PAAccount` to switch to the existing account.

After creating the account, requesting a certificate actually works and i receive the certificate files!
When i then switch accounts to one of the earlier created, they are deactivated.

Even freshly created accounts appear as `deactivated` when i use `Get-PAAccount -Refresh` or use `Set-PAAccount` to change accounts!

Here's an isolated example to showcase the issue. Unfortunately i had to redact a lot of stuff, since i didn't know for certain what is sensitive and what not. If you're missing something specific, please tell me!

```PowerShell
$vmHost = "vmhostname.example.com"
PS C:\Users\USERNAME> New-PAAccount -AcceptTOS -Contact "contactmail@example.com" -ID $vmHost -Verbose -Debug
DEBUG: Refreshing valid accounts
DEBUG: Loading PAAccount list from disk
DEBUG: Loading PAAccount list from disk
DEBUG: Creating new ec-256 account with contact: mailto:contactmail@example.com
DEBUG: Creating new EC 256 key
DEBUG: ACME Header:
{
  "url": "https://acme.example.com/certACME/WebServer/new-account",
  "jwk": {
    "crv": "P-256",
    "kty": "EC",
    "x": "REDACTED",
    "y": "REDACTED"
  },
  "nonce": "REDACTED-nonceA",
  "alg": "ES256"
}
DEBUG: ACME Payload:
{"termsOfServiceAgreed":true,"contact":["mailto:contactmail@example.com"]}
DEBUG: Signing message using EC with SHA256
DEBUG: POST https://acme.example.com/certACME/WebServer/new-account
{"payload":"REDACTED","protected":"REDACTED","signature":"REDACTED"}
DEBUG: ACME Response:
{"accountId":"REDACTED-AccountID","status":"Valid","orders":"https://acme.example.com/certACME/WebServer/account/REDACTED-AccountID/orders","contact":["mailto:contactmail@example.com"],"termsOfServiceAgreed":true}
DEBUG: Updated nonce: REDACTED-nonceB

id                   status          contact                             alg        KeyLength
--                   ------          -------                             ---        ---------
vmhostname.example.… Valid           {mailto:contactmail@example.com}       ES256      ec-256

PS C:\Users\USERNAME> Get-PAAccount -ID $vmHost -Debug -Verbose
DEBUG: Loading PAAccount vmhostname.example.com from disk

id                   status          contact                             alg        KeyLength
--                   ------          -------                             ---        ---------
vmhostname.example.… Valid           {mailto:contactmail@example.com}       ES256      ec-256

PS C:\Users\USERNAME> Get-PAAccount -ID $vmHost -Debug -Verbose -Refresh
DEBUG: Loading PAAccount vmhostname.example.com from disk
DEBUG: Refreshing account vmhostname.example.com
DEBUG: Refreshing account vmhostname.example.com
DEBUG: ACME Header:
{
  "url": "https://acme.example.com/certACME/WebServer/account/REDACTED-AccountID",
  "kid": "https://acme.example.com/certACME/WebServer/account/REDACTED-AccountID",
  "nonce": "REDACTED-nonceB",
  "alg": "ES256"
}
DEBUG: ACME Payload: (empty)
DEBUG: Signing message using EC with SHA256
DEBUG: POST https://acme.example.com/certACME/WebServer/account/REDACTED-AccountID
{"payload":"","protected":"REDACTED","signature":"REDACTED"}
DEBUG: ACME Response:
{"status":"deactivated"}
DEBUG: Updated nonce: REDACTED-nonceC
DEBUG: Loading PAAccount vmhostname.example.com from disk

id                       status          contact    alg        KeyLength
--                       ------          -------    ---        ---------
vmhostname.example.com.… deactivated                ES256      ec-256
```

Could this be related to Posh-ACME's code? Since according to RFC 8555, only the client should be able to deactivate an account?
If not, what could be causing this issue? Does Secardeo have to look into it on their side?

**Comment by rmbolger** [comment]
Hi @eren-kemer, thanks for reaching out. From the debug output you've posted, there's nothing that would be deactivating the account from the Posh-ACME side. Account refreshes are a standard ACME [POST-as-GET](https://www.rfc-editor.org/rfc/rfc8555.html#section-6.3) operation. They don't change the state of things. This seems like some sort of automated account deactivation/timeout feature in certACME.

**Comment by rmbolger** [comment]
As I'm thinking more about this, there have been some public CAs that have poor RFC8555 compliance and historically haven't handled POST-as-GET requests for account resources very well. None of them outright deactivated an account, they just threw an error.

To workaround those problematic CAs, I added an option with `Set-PAServer` that tells the module to use an alternative method for refreshing the status of the account that uses the `newAccount` endpoint. Try this out and see if the account is still deactivated:

```powershell
Set-PAServer -UseAltAccountRefresh
New-PAAccount -AcceptTOS -Contact "contactmail@example.com" -ID $vmHost -Verbose
Get-PAAccount -ID $vmHost -Refresh
```

**Comment by eren-kemer** [comment]
This seems to have worked in my small testcase. I will test it in the actual script and get back with the result!
Thank you!

Is this a non-compliance with RFC8555 from certACME, or just a poor implementation?

**Comment by rmbolger** [comment]
I'd guess this is a bug in the certACME implementation as opposed to something working as intended. It seems to be treating the empty ACME payload in the POST-as-GET request as a request to deactivate the account instead of just a request to retrieve the account details. It's like they're accidentally using normal REST semantics instead of ACME specific semantics where a POST with an empty payload implies you want to delete the resource you're referencing.

Account deactivation is documented in [section 7.3.6](https://www.rfc-editor.org/rfc/rfc8555.html#section-7.3.6) of the spec and should require passing `"status": "deactivated"` in the payload. But Posh-ACME only does that if you run `Set-PAAccount -Deactivate`.

So basically they need to re-read sections 6.3 for POST-as-GET and 7.3.6 for account deactivation and fix their implementation.

**Comment by eren-kemer** [comment]
Hey, so i was able to fix the issue with accounts being deactivated on -Refresh.

But i noticed that whenever I get to actually requesting a certificate, UseAltAccountRefresh is set to false by Posh-.ACME?
Is that supposed to happen, even if Set-PAServer was used with -UseAltAccountRefresh? I think that is still causing issues for me with the several accounts.

```Powershell
VERBOSE: Updating directory info from https://<hostname>/certACME/WebServer
DEBUG: Requesting nonce from https://<hostname>/certACME/WebServer/new-nonce
DEBUG: Setting UseAltAccountRefresh value to False
DEBUG: Saving PAServer to disk
DEBUG: Loading PAServer list from disk
DEBUG: Enabling cert validation
DEBUG: Loading PAAccount <paaccountname> from disk
DEBUG: Loading PAOrder list from disk
```

**Comment by eren-kemer** [comment]
Also, is there a way to handle a CA Server deactivating accounts themselves? Currently the script recognizes that the account had deactivated instead of valid, but throws an Exception instead of deactivating it locally?

```Powershell
DEBUG: Response Code 409, Body:
{"type":"urn:ietf:params:acme:error:malformed","detail":"The account used in this request did not have the expected status 'Valid' but had 'Deactivated'.","status":405}
OperationStopped: C:\Program Files\WindowsPowerShell\Modules\Posh-ACME\4.16.0\Private\Invoke-ACME.ps1:174
Line |
 174 |          throw [AcmeException]::new($acmeError.detail,$acmeError)
     |          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | The account used in this request did not have the expected status 'Valid' but had 'Deactivated'.
```

**Comment by rmbolger** [comment]
> But i noticed that whenever I get to actually requesting a certificate, UseAltAccountRefresh is set to false by Posh-ACME?
Is that supposed to happen, even if Set-PAServer was used with -UseAltAccountRefresh?

No, that shouldn't happen and sounds like a bug if it's reproducible. Can you post the series of commands that trigger the behavior?

> Also, is there a way to handle a CA Server deactivating accounts themselves (even though it deviates from the RFC)? Currently the script recognizes that the account had deactivated instead of valid, but throws an Exception instead of deactivating it locally?

In that output, the module is just forwarding the error reported by the ACME server which is basically that you tried to perform some action using a deactivated account. Normally the module wouldn't/shouldn't be letting you do whatever it was using a deactivated account. But in this case, the metadata associated with the locally cached account didn't indicate it was disabled probably because the ACME server had disabled it without our knowledge. So the workaround would basically be to do an explicit account refresh before every new action.

**Comment by eren-kemer** [comment]
> No, that shouldn't happen and sounds like a bug if it's reproducible. Can you post the series of commands that trigger the behavior?

This is the code i'm essentially running:

```Powershell
If ((Get-PAServer).Name -ne $Settings.acmeurl)  {
        Set-PAServer $Settings.acmeURL -UseAltAccountRefresh        
    }
$existingAccounts = Get-PAAccount -List

if ($null -ne $existingAccounts -and ($existingAccounts.id.Contains($vmHost.Name))) {
        Set-PAAccount $vmHost.Name -Verbose
        if ((Get-PAAccount -).status -eq "deactivated") {
            Write-Host "Current Hosts ACME Account is deactivated. Cancelling Cert Request!"
            return $false
        }
    }
else {
        New-PAAccount -AcceptTOS -Contact $Settings.contactmail -ID $vmHost.Name -Verbose -Debug -Force
}

$certificate = New-PACertificate -CSRPath "$($vmHost.Name).csr" -Contact $Settings.contactmail -Plugin WebSelfHost -Verbose -Debug

```

This part was mostly for testing, before you provided the -UseAltAccountRefresh workaround. So it could be ignored:
```Powershell
if ((Get-PAAccount -).status -eq "deactivated") {
        Write-Host "Current Hosts ACME Account is deactivated. Cancelling Cert Request!"
        return $false
```

This is the steps i took to reproduce it:

```Powershell
PS> (Get-PAServer).UseAltAccountRefresh
False
PS> Set-PAServer <acmeurl> -UseAltAccountRefresh
PS> (Get-PAServer).UseAltAccountRefresh
True
PS> .\poshacme.ps1 -Debug
VERBOSE: Updating directory info from <acmeurl>/certACME/WebServer
DEBUG: Requesting nonce from https://<acmeurl>/certACME/WebServer/new-nonce
DEBUG: Setting UseAltAccountRefresh value to False
DEBUG: Saving PAServer to disk
DEBUG: Loading PAServer list from disk
DEBUG: Enabling cert validation
DEBUG: Loading PAAccount <paaccountname> from disk
DEBUG: Loading PAOrder list from disk
[...]
PS> (Get-PAServer).UseAltAccountRefresh
False
```


If you need any more information, please tell me what is required!

**Comment by rmbolger** [comment]
Thanks for the detailed repro. It led me straight to a rather embarrassing bug where any of the switches on a PAServer object that have been set to `$true` get reset to their `$false` default on subsequent calls to `Set-PAServer` that don't explicitly set them to `$true` again. `New-PACertificate` calls it under the hood with no parameters to refresh the nonce from the ACME server which is what triggered it in your script.

I'll try and have a fix out in the next day or two.
